site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+	// +optional
+	// +kubebuilder:default={type: "Exact"}
+	Method *GRPCMethodMatch `json:"path,omitempty"`
+
+	// Headers specifies gRPC request header matchers. Multiple match values are
+	// ANDed together, meaning, a request must match all the specified headers
+	// to select the route.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []GRPCHeaderMatch `json:"headers,omitempty"`
+}
+
+// GRPCPathMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method..
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact)
+	//
+	// Support: Custom (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Service *string `json:"value,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Method *string `json:"value,omitempty"`
+}
+
+// MethodMatchType specifies the semantics of how gRPC methods and services should be compared.
+// Valid MethodMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
+//
+// Exact paths must be syntactically valid:
+//
+// - Must not contain `/` character
+//
+// +kubebuilder:validation:Enum=Exact;PathPrefix;RegularExpression
+// +kubebuilder:validation:Enum=Exact;RegularExpression
+type GRPCMethodMatchType string
+
+const (
+	// Matches the service and/or method exactly and with case sensitivity.
+	PathMatchExact PathMatchType = "Exact"
+
+	// Matches if the service and/or method matches the given regular expression with
+	// case sensitivity.
+	//
+	// Since `"RegularExpression"` has custom conformance, implementations
+	// can support POSIX, PCRE, RE2 or any other regular expression dialect.
+	// Please read the implementation's documentation to determine the supported
+	// dialect.
+	PathMatchRegularExpression PathMatchType = "RegularExpression"
+)
+
+// GRPCHeaderMatch describes how to select a gRPC route by matching gRPC request
+// headers.
+type GRPCHeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name is the name of the gRPC Header to be matched.
+	//
+	// If multiple entries specify equivalent header names, only the first
+	// entry with an equivalent name MUST be considered for a match. Subsequent
+	// entries with an equivalent header name MUST be ignored. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered
+	// equivalent.
+	Name GRPCHeaderName `json:"name"`
```

Comment:
Isn't a gRPC header the same as a HTTP header? My knowledge in this area is weak so please bear with me.

Reply:
This is true, roughly speaking. More details [here](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md).

Reply:
It may make sense to call this "HeaderName" instead to avoid unnecessary confusion. The "GRPC" prefix doesn't bring any additional value but can cause confusion here.